halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1351,6 +1466,17 @@ impl<F: Field> ConstraintSystem<F> {
         tmp
     }
 
+    /// Allocate a new challenge in given phase
+    pub fn challenge_in(&mut self, phase: Phase) -> Challenge {
```

Comment:
Per my earlier comment, the challenge should not be interpreted as allocated "in" `phase`, but instead between `phase` and `phase + 1`. I don't have any great succinct naming suggestions right now, so maybe we should just be verbose:
```suggestion
    /// Requests a challenge that is usable after the given phase.
    pub fn challenge_usable_after(&mut self, phase: Phase) -> Challenge {
```
This is also where I think using type markers instead of an enum would give a better public API, so the `Phase` enum could be entirely private:
```suggestion
    /// Requests a challenge that is usable after the given phase.
    pub fn challenge_usable_after<P: Phase>(&mut self, phase: P) -> Challenge {
```
and then
```rust
trait SealedPhase {
    fn to_enum(self) -> PhaseEnum;
}
pub trait Phase: SealedPhase {}

enum FirstPhase {}
impl SealedPhase for FirstPhase {
    fn to_enum(self) -> PhaseEnum {
        PhaseEnum::First
    }
}
impl Phase for FirstPhase {}
```
`Challenge` could still store `PhaseEnum` internally to stop the `Phase` trait bound from propagating further than this API (which it has no need to, as phases don't need to appear anywhere else).

Reply:
While I was trying sealed pattern for this, I found that `PhaseEnum` might still need to be public because `Advice` and `Any::Advice` has field of it, and some failure test cases need to create `VirtualCell` to be matched, which then needs to specify advice column's phase (even it doesn't use multi-phase feature).

Perhaps we might need to create `VirtualCell` in a way that it doesn't need `PhaseEnum`?

Reply:
Yeah, this should be handled by changing the way `VirtualCell` works. If the dev tooling is the only thing placing this requirement on the public API, it's the dev tooling we should change.

Reply:
I suspect the easiest change here is to add a `VirtualColumn` that allows setting the phase for advice, and then have an `impl From<Any> for VirtualColumn` etc. We could also keep the existing `From` impls, so fixed and instance checks wouldn't have to change.

Reply:
I see, it makes sense to me, will try to add `VirtualColumn`.

The other issue I have just encountered is in `Expression::evaluate`, if we have `PhaseEnum` as private enum, what should we pass to the function to evaluate advice column (we can't pass `PhaseEnum` because `Expression::evaluate` is public)?

Reply:
Yes, I think introducing `*Query` structs is the way to go here; then future changes like this will be simpler, the evaluate interface will be simpler, and we can control what is exposed.

Reply:
I see, I will try to work on these `*Query` structs first in a separate PR, and then get back to resolve this one.